File: tlv/src/main/java/dev/keiji/tlv/BerTlvDecoder.java
package dev.keiji.tlv;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InvalidObjectException;
import java.io.StreamCorruptedException;
import java.math.BigInteger;

public class BerTlvDecoder {
    private static int MASK_MSB_BITS = 0b100_00000;
    private static int MASK_PC_BITS = 0b00_1_00000;
    private static int MASK_TAG_BITS = 0b00_0_11111;
    private static byte MASK_TAG_FULL_BITS = 0b00_0_11111;

    private static int PC_PRIMITIVE = 0b0;
    private static int PC_CONSTRUCTED = 0b1;

    private static int MAX_LENGTH_FILED_LENGTH = 126;

    public interface Callback {
        void onLargeItemDetected(
                byte[] tag,
                BigInteger length,
                InputStream inputStream
        );

        void onItemDetected(
                byte[] tag,
                byte[] data
        );
    }

    public static void readFrom(
            InputStream inputStream,
            Callback callback
    ) throws IOException {
        if (inputStream == null) {
            throw new IllegalArgumentException("inputStream must not be null.");
        }
        if (callback == null) {
            throw new IllegalArgumentException("callback must not be null.");
        }

        while (true) {
            byte[] tag = readTag(inputStream);
            if (tag == null) {
                break;
            }

            BigInteger length = readLength(inputStream);

            boolean isLargeItem = length.bitLength() > (Integer.SIZE - 1);
            if (!isLargeItem) {
                byte[] data = readData(inputStream, length);
                callback.onItemDetected(tag, data);
            } else {
                callback.onLargeItemDetected(tag, length, inputStream);
            }
        }

    }

    static byte[] readTag(InputStream inputStream) throws IOException {
        ByteArrayOutputStream tagStream = new ByteArrayOutputStream();

        while (true) {
            int b = inputStream.read();
            if (b < 0) {
                return null;
            }

            byte tagBits = (byte) (b & MASK_TAG_BITS);

            int tagSize = tagStream.size();

            if (tagSize == 0) {
                if (tagBits != MASK_TAG_FULL_BITS) {
                    return new byte[]{(byte) b};
                }
                tagStream.write(new byte[]{(byte) b});
            } else {
                tagStream.write(new byte[]{(byte) b});
                if ((b & MASK_MSB_BITS) != 0) {
                    continue;
                }
                return tagStream.toByteArray();
            }
        }
    }

    static BigInteger readLength(InputStream inputStream) throws IOException {
        int b = inputStream.read();
        if (b < 0) {
            throw new StreamCorruptedException();
        }

        boolean longDef = (b & MASK_MSB_BITS) != 0;

        if (!longDef) {
            return BigInteger.valueOf(b);
        } else {
            int fieldLength = (b ^ MASK_MSB_BITS);
            if (fieldLength > MAX_LENGTH_FILED_LENGTH) {
                throw new InvalidObjectException("Long Definite length must not be grater 126 bytes.");
            }

            byte[] lengthBytes = new byte[fieldLength];

            int offset = 0;

            while (true) {
                int readLength = inputStream.read(lengthBytes, offset, (fieldLength - offset));
                if (readLength < 0) {
                    throw new StreamCorruptedException();
                } else if (readLength < (fieldLength - offset)) {
                    offset += readLength;
                } else {
                    break;
                }
            }
            return new BigInteger(+1, lengthBytes);
        }
    }

    static byte[] readData(InputStream inputStream, BigInteger length) throws IOException {
        int dataLength = length.intValue();
        byte[] data = new byte[dataLength];
        int offset = 0;

        while (true) {
            int readLength = inputStream.read(data, offset, (dataLength - offset));
            if (readLength < 0) {
                throw new StreamCorruptedException();
            } else if (readLength < (dataLength - offset)) {
                offset += readLength;
            } else {
                return data;
            }
        }
    }
}

File: tlv/src/main/java/dev/keiji/tlv/BerTlvDecoder.kt
/*
 * Copyright (C) 2022 ARIYAMA Keiji
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package dev.keiji.tlv

import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.io.InvalidObjectException
import java.io.StreamCorruptedException
import java.math.BigInteger

/**
 * BER-TLV Decoder.
 */
class BerTlvDecoder {
    companion object {
        private const val MASK_MSB_BITS = 0b100_00000
        private const val MASK_PC_BITS = 0b00_1_00000
        private const val MASK_TAG_BITS = 0b00_0_11111
        private const val MASK_TAG_FULL_BITS: Byte = 0b00_0_11111

        private const val PC_PRIMITIVE = 0b0
        private const val PC_CONSTRUCTED = 0b1

        private const val MAX_LENGTH_FILED_LENGTH = 126

        /**
         * Read TLV-item(s) from the InputStream.
         *
         * @param inputStream
         * @param callback Callback to receive event about detect TLV-item
         */
        fun readFrom(
            inputStream: InputStream,
            callback: Callback,
        ) {
            while (true) {
                val tag = readTagFieldBytes(inputStream)
                tag ?: break

                val length = readLength(inputStream)
                if (length == null) {
                    callback.onUnknownLengthItemDetected(tag, inputStream)
                    continue
                }

                // If tag:0x00 length:0x00 have been detected the parser decide stream have been finished.
                if (byteArrayOf(0x00).contentEquals(tag) && BigInteger.ZERO.equals(length)) {
                    break
                }

                val isLargeItem = length.bitLength() > (Integer.SIZE - 1)
                if (!isLargeItem) {
                    val value = readValueFieldBytes(inputStream, length)
                    callback.onItemDetected(tag, value)
                } else {
                    callback.onLargeItemDetected(tag, length, inputStream)
                }
            }
        }

        fun readValueFieldBytes(
            inputStream: InputStream,
            length: BigInteger,
        ): ByteArray {
            val dataLength = length.toInt()
            val data = ByteArray(dataLength)
            var offset = 0

            while (true) {
                val readLength = inputStream.read(data, offset, (dataLength - offset))
                if (readLength < 0) {
                    throw StreamCorruptedException()
                } else if (readLength < (dataLength - offset)) {
                    offset += readLength
                } else {
                    return data
                }
            }
        }

        fun readTagFieldBytes(inputStream: InputStream): ByteArray? {
            val tagStream = ByteArrayOutputStream()

            while (true) {
                val b = inputStream.read()
                if (b < 0) {
                    return null
                }

                val tagBits = (b and MASK_TAG_BITS).toByte()
                val tagSize = tagStream.size()

                if (tagSize == 0) {
                    if (tagBits != MASK_TAG_FULL_BITS) {
                        return byteArrayOf(b.toByte())
                    }
                    tagStream.write(byteArrayOf(b.toByte()))
                } else {
                    tagStream.write(byteArrayOf(b.toByte()))
                    if (b and MASK_MSB_BITS == 0) {
                        return tagStream.toByteArray()
                    }
                }
            }
        }

        fun readLengthFieldBytes(inputStream: InputStream): ByteArray? {
            val b = inputStream.read()
            if (b < 0) {
                throw StreamCorruptedException()
            }

            // Length undefined
            if (b == MASK_MSB_BITS) {
                return null
            }

            val longDef = b and MASK_MSB_BITS != 0

            if (!longDef) {
                return byteArrayOf(b.toByte())
            }

            val fieldLength = b xor MASK_MSB_BITS
            if (fieldLength > MAX_LENGTH_FILED_LENGTH) {
                throw InvalidObjectException("Long Definite length must not be grater 126 bytes.")
            }

            val lengthFieldSize = 1 + fieldLength
            val lengthFieldBytes = ByteArray(lengthFieldSize).also {
                it[0] = b.toByte()
            }

            var offset = 1
            while (true) {
                val readLength =
                    inputStream.read(lengthFieldBytes, offset, (lengthFieldSize - offset))
                if (readLength < 0) {
                    throw StreamCorruptedException()
                } else if (readLength < (fieldLength - offset)) {
                    offset += readLength
                } else {
                    break
                }
            }
            return lengthFieldBytes
        }

        fun readLength(inputStream: InputStream): BigInteger? {
            val lengthBytes = readLengthFieldBytes(inputStream) ?: return null
            val lengthFieldSize = lengthBytes.size

            return if (lengthFieldSize <= 1) {
                BigInteger(+1, lengthBytes)
            } else {
                BigInteger(+1, lengthBytes.copyOfRange(1, lengthFieldSize))
            }
        }
    }

    /**
     * An interface to receive event about detect TLV-item on InputStream.
     * The Callback is set with BerTlvDecoder.readFrom method.
     */
    interface Callback {

        /**
         * Called when a TLV-item has been detected.
         *
         * @param tag
         * @param value
         */
        fun onItemDetected(tag: ByteArray, value: ByteArray)

        /**
         * Called when an unknown length TLV-item has been detected.
         *
         * @param tag
         * @param inputStream
         */
        fun onUnknownLengthItemDetected(
            tag: ByteArray,
            inputStream: InputStream
        ) {
        }

        /**
         * Called when a large TLV-item has been detected.
         *
         * Note: BER-TLV length can be up to 126 bytes.
         *
         * @param tag
         * @param length
         * @param inputStream
         */
        fun onLargeItemDetected(
            tag: ByteArray,
            length: BigInteger,
            inputStream: InputStream
        ) {
        }
    }
}
